src/documentation/DocumentationPreviewEditor.ts

```diff
@@ -0,0 +1,166 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the VS Code Swift open source project
+//
+// Copyright (c) 2024 the VS Code Swift project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of VS Code Swift project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import * as vscode from "vscode";
+import * as fs from "fs/promises";
+import * as path from "path";
+import { RenderNode, WebviewContent, WebviewMessage } from "./webview/WebviewMessage";
+import { WorkspaceContext } from "../WorkspaceContext";
+import { ConvertDocumentationRequest } from "../sourcekit-lsp/extensions/ConvertDocumentationRequest";
+
+export enum PreviewEditorConstant {
+    VIEW_TYPE = "swift.previewDocumentationEditor",
+    TITLE = "Preview Swift Documentation",
+}
+
+export class DocumentationPreviewEditor implements vscode.Disposable {
+    private readonly webviewPanel: vscode.WebviewPanel;
+    private subscriptions: vscode.Disposable[] = [];
+
+    private disposeEmitter = new vscode.EventEmitter<void>();
+    private renderEmitter = new vscode.EventEmitter<void>();
+    private updateContentEmitter = new vscode.EventEmitter<WebviewContent>();
+
+    constructor(
+        private readonly extension: vscode.ExtensionContext,
+        private readonly context: WorkspaceContext
+    ) {
+        const swiftDoccRenderPath = this.extension.asAbsolutePath("assets/swift-docc-render");
+        // Create and hook up events for the WebviewPanel
+        this.webviewPanel = vscode.window.createWebviewPanel(
+            PreviewEditorConstant.VIEW_TYPE,
+            PreviewEditorConstant.TITLE,
+            { viewColumn: vscode.ViewColumn.Beside, preserveFocus: true },
+            {
+                enableScripts: true,
+                localResourceRoots: [
+                    vscode.Uri.file(this.extension.asAbsolutePath("assets/documentation-webview")),
+                    vscode.Uri.file(swiftDoccRenderPath),
+                    ...context.folders.map(f => f.folder),
+                ],
+            }
+        );
+        const webviewBaseURI = this.webviewPanel.webview.asWebviewUri(
+            vscode.Uri.file(swiftDoccRenderPath)
+        );
+        const scriptURI = this.webviewPanel.webview.asWebviewUri(
+            vscode.Uri.file(this.extension.asAbsolutePath("assets/documentation-webview/index.js"))
```

Comment:
nit: path.join like below

Reply:
also is the build placing this script under assets? Is there a reason we cannot place this and swift-docc-renderer under dist like rest of build artifacts so we only have the one build folder.

Reply:
Done.

The build does place the script into assets because it needs to be bundled with esbuild to work. The TypeScript compiler doesn't get rid of imports and those don't work in webviews.

Reply:
ok, but to confirm, esbuild is putting this under dist/ when building?

Reply:
No, it doesn't, unfortunately. This is due to some weirdness in our build where we use `tsc` for local development and `esbuild` for production. I will fix this in a subsequent PR.